File: src/main/java/ru/arc/hooks/citizens/CitizensHook.java
package ru.arc.hooks.citizens;

import ru.arc.ARC;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import lombok.extern.slf4j.Slf4j;
import net.citizensnpcs.api.CitizensAPI;
import net.citizensnpcs.api.npc.NPC;
import net.citizensnpcs.api.trait.trait.Equipment;
import net.citizensnpcs.trait.HologramTrait;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.EntityType;
import org.bukkit.inventory.ItemStack;

import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import static ru.arc.util.Logging.debug;
import static ru.arc.util.Logging.warn;

@Slf4j
public class CitizensHook {

    private static CitizensListener listener;

    public void registerListeners() {
        if (listener != null) return;
        listener = new CitizensListener();
        Bukkit.getPluginManager().registerEvents(listener, ARC.plugin);
    }

    public int createNpc(String name, Location location) {
        NPC npc = CitizensAPI.getNPCRegistry().createNPC(EntityType.PLAYER, name);
        //npc.setName(name);
        npc.spawn(location);
        return npc.getId();
    }

    public void deleteWithNames(Set<String> npcNames) {
        CitizensAPI.getNPCRegistry().forEach(npc -> {
            if (npcNames.contains(npc.getName())) {
                npc.destroy();
            }
        });
    }

    public record HologramLine(String text, int ticks) {
    }

    public record InsertedHologramLine(int line, long expireTime) {
    }

    Cache<Integer, ConcurrentLinkedDeque<InsertedHologramLine>> linesCache = CacheBuilder.newBuilder()
            .expireAfterAccess(10, TimeUnit.MINUTES)
            .build();

    public void addChatBubble(int id, List<HologramLine> lineList) {
        try {
            NPC npc = CitizensAPI.getNPCRegistry().getById(id);
            if (npc == null) {
                warn("NPC {} is null", id);
                return;
            }
            HologramTrait trait = npc.getOrAddTrait(HologramTrait.class);
            var lineCache = linesCache.get(id, ConcurrentLinkedDeque::new);
            for (InsertedHologramLine line : lineCache.reversed()) {
                trait.removeLine(line.line);
            }
            lineCache.clear();

            AtomicInteger initialSize = new AtomicInteger(trait.getLines().size());
            lineList.reversed().forEach(line -> {
                trait.addTemporaryLine(line.text, line.ticks);
                lineCache.add(new InsertedHologramLine(initialSize.getAndIncrement(), System.currentTimeMillis() + line.ticks * 50L));
            });
        } catch (Exception e) {
            warn("Error adding hologram lines", e);
        }
    }

    public void lookClose(int id) {
        try {
            ARC.trySeverCommand("npc lookclose --id " + id);
        } catch (Exception e) {
            debug("Error looking close", e);
        }
    }

    public void setSkin(int id, String link) {
        ARC.trySeverCommand("npc skin --url " + link + " --id " + id);
    }

    public void deleteNpc(int id) {
        NPC npc = CitizensAPI.getNPCRegistry().getById(id);
        if (npc == null) return;
        npc.destroy();
    }

    public void faceNpc(int id, Location location) {
        NPC npc = CitizensAPI.getNPCRegistry().getById(id);
        if (npc == null) return;
        npc.faceLocation(location);
    }

    public void animateNpc(int id, Animation animation) {
        try {
            ARC.trySeverCommand("npc panimate " + animation.name() + " --id " + id);
        } catch (Exception e) {
            debug("Error animating npc", e);
        }
    }
    @SuppressWarnings("deprecation")
    public void setMainHand(int id, ItemStack stack) {
        try {
            NPC npc = CitizensAPI.getNPCRegistry().getById(id);
            if (npc == null) return;
            npc.getTrait(Equipment.class).set(Equipment.EquipmentSlot.HAND, stack);
        } catch (Exception e) {
            debug("Error setting main hand", e);
        }
    }

    public enum Animation {
        ARM_SWING, SIT, STOP_SITTING
    }
}

File: src/main/kotlin/ru/arc/hooks/citizens/CitizensHook.kt
package ru.arc.hooks.citizens

import net.citizensnpcs.api.CitizensAPI
import net.citizensnpcs.api.trait.trait.Equipment
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.entity.EntityType
import org.bukkit.entity.LivingEntity
import org.bukkit.inventory.ItemStack
import ru.arc.ARC
import ru.arc.util.Logging.debug
import ru.arc.util.Logging.warn

class CitizensHook : AutoCloseable {

    data class HologramLine(val text: String, val ticks: Int)

    enum class Animation { ARM_SWING, SIT, STOP_SITTING }

    data class NearbyNpc(
        val id: Int,
        val name: String,
        val location: Location,
    )

    private var closed = false

    @Synchronized
    override fun close() {
        if (closed) return
        closed = true
    }

    fun createNpc(name: String, location: Location): Int {
        return try {
            val npc = CitizensAPI.getNPCRegistry().createNPC(EntityType.PLAYER, name)
            npc.spawn(location)
            debug("[autobuild] Created NPC id={} name={} at {}", npc.id, name, location)
            npc.id
        } catch (e: Exception) {
            warn("Failed to create NPC {} at {}", name, location, e)
            -1
        }
    }

    fun findNearestNpc(
        origin: Location,
        expectedName: String,
        maxDistance: Double,
    ): NearbyNpc? {
        val world = origin.world ?: return null
        return CitizensAPI.getNPCRegistry()
            .asSequence()
            .filter { it.isSpawned && ArcNpcHologramModule.matchesName(it, expectedName) }
            .mapNotNull { npc ->
                val location = npc.entity?.location ?: return@mapNotNull null
                if (location.world?.uid != world.uid) return@mapNotNull null
                val distanceSquared = location.distanceSquared(origin)
                if (distanceSquared > maxDistance * maxDistance) return@mapNotNull null
                Triple(npc, location, distanceSquared)
            }
            .minByOrNull { it.third }
            ?.let { (npc, location) -> NearbyNpc(npc.id, npc.name, location.clone()) }
    }

    fun livingEntity(id: Int): LivingEntity? =
        CitizensAPI.getNPCRegistry().getById(id)?.entity as? LivingEntity

    fun deleteWithNames(npcNames: Set<String>) {
        CitizensAPI.getNPCRegistry().forEach { npc ->
            if (npcNames.contains(npc.name)) npc.destroy()
        }
    }

    fun addChatBubble(id: Int, lineList: List<HologramLine>) {
        try {
            val ttlTicks = lineList.maxOfOrNull(HologramLine::ticks) ?: 1
            ArcNpcHologramModule.showTemporaryBubble(id, lineList.map(HologramLine::text), ttlTicks)
        } catch (e: Exception) {
            warn("Error adding hologram lines", e)
        }
    }

    fun lookClose(id: Int) {
        try {
            ARC.trySeverCommand("npc lookclose --id $id")
        } catch (e: Exception) {
            debug("Error looking close", e)
        }
    }

    fun setSkin(id: Int, link: String) {
        ARC.trySeverCommand("npc skin --url $link --id $id")
    }

    fun deleteNpc(id: Int) {
        CitizensAPI.getNPCRegistry().getById(id)?.destroy()
    }

    fun faceNpc(id: Int, location: Location) {
        CitizensAPI.getNPCRegistry().getById(id)?.faceLocation(location)
    }

    fun animateNpc(id: Int, animation: Animation) {
        try {
            ARC.trySeverCommand("npc panimate ${animation.name} --id $id")
        } catch (e: Exception) {
            debug("Error animating npc", e)
        }
    }

    @Suppress("DEPRECATION")
    fun setMainHand(id: Int, stack: ItemStack) {
        try {
            val npc = CitizensAPI.getNPCRegistry().getById(id) ?: return
            npc.getTrait(Equipment::class.java).set(Equipment.EquipmentSlot.HAND, stack)
        } catch (e: Exception) {
            debug("Error setting main hand", e)
        }
    }
}
